

<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>RNA World Hypothesis &#8212; ASTROBIOLOGY 200</title>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" integrity="sha384-KA6wR/X5RY4zFAHpv/CnoG2UW1uogYfdnP67Uv7eULvTveboZJg0qUpmJZb5VqzN" crossorigin="anonymous">
    <link href="../../../_static/css/index.css" rel="stylesheet">
    <link rel="stylesheet" href="../../../_static/sphinx-book-theme.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/jupyter-sphinx.css" />
    <script id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/sphinx-book-theme.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/language_data.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/mystnb.js"></script>
    <script src="../../../_static/sphinx-book-theme.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script async="async" src="../../../_static/thebelab.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Habitability of Inner vs. Outer Solar System" href="../../08-habitability-our-system.html" />
    <link rel="prev" title="Complex Life" href="../1/complex-life.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="docsearch:language" content="en">



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../../../index.html">
  
  <img src="../../../_static/astro_logo.png" class="logo" alt="logo">
  
  
  <h1 class="site-logo" id="site-title">ASTROBIOLOGY 200</h1>
  
</a>
</div>

<form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>

<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
  
  <ul class="nav sidenav_l1">
  <li class="">
    <a href="../../../intro.html">ASTROBIOLOGY 200</a>
  </li>
  <li class="">
    <a href="../../LICENSE.html">License for this book</a>
  </li>
<li class="navbar-special">
<p class="margin-caption">I. THE CONTEXT FOR LIFE IN THE UNIVERSE</p>
</li>
  <li class="">
    <a href="../../01-introduction.html">Introduction</a>
  <ul class="nav sidenav_l2">
    <li class="">
      <a href="../../01/1/what-is-astrobiology.html">What is astrobiology?</a>
    </li>
    <li class="">
      <a href="../../01/2/the-ways-of-science.html">The-ways-of-science</a>
    </li>
    <li class="">
      <a href="../../01/3/from_ptolemy_to_kepler.html">From Ptolemy to Kepler</a>
    </li>
    <li class="">
      <a href="../../01/4/distances_in_the_universe.html">Distances In The Universe</a>
    </li>
  </ul>
  </li>
  <li class="">
    <a href="../../02-scales-in-astronomy.html">Scales of the Universe</a>
  <ul class="nav sidenav_l2">
    <li class="">
      <a href="../../02/Radioactivity.html">Radioactive decay</a>
    </li>
    <li class="">
      <a href="../../02/DopplerShift.html">Doppler Shift</a>
    </li>
  </ul>
  </li>
  <li class="">
    <a href="../../03-intro-to-geology.html">Intro to Geology</a>
  </li>
  <li class="">
    <a href="../../04-earth-habitability.html">Earth’s Habitability</a>
  </li>
<li class="navbar-special">
<p class="margin-caption">II. LIFE ON EARTH</p>
</li>
  <li class="">
    <a href="../../05-biology-of-life.html">The Biology of Life</a>
  <ul class="nav sidenav_l2">
    <li class="">
      <a href="../../05/1/what-is-life.html">What is Life?</a>
    </li>
    <li class="">
      <a href="../../05/2/DNA-discovery.html">The DNA molecule</a>
    </li>
  </ul>
  </li>
  <li class="">
    <a href="../../06-life-limits.html">Life's Limits</a>
  <ul class="nav sidenav_l2">
    <li class="">
      <a href="../../06/1/early-earth.html">What is Life?</a>
    </li>
  </ul>
  </li>
  <li class="active">
    <a href="../../07-origin-of-life.html">The Origin of Life</a>
  <ul class="nav sidenav_l2">
    <li class="">
      <a href="../1/complex-life.html">History of Complex Life?</a>
    </li>
    <li class="active">
      <a href="">RNA World Hypothesis</a>
    </li>
  </ul>
  </li>
<li class="navbar-special">
<p class="margin-caption">III. LIFE IN THE SOLAR SYSTEM</p>
</li>
  <li class="">
    <a href="../../08-habitability-our-system.html">Habitability in the Solar System</a>
  </li>
  <li class="">
    <a href="../../09-geology-of-mars.html">Mars</a>
  </li>
  <li class="">
    <a href="../../10-moons-of-giants.html">The Moons of the Giants</a>
  </li>
<li class="navbar-special">
<p class="margin-caption">IV. LIFE AMONGST THE STARS</p>
</li>
  <li class="">
    <a href="../../11-seti-drake.html">SETI and Drake's Equation</a>
  </li>
  <li class="">
    <a href="../../12-future-of-earth.html">Future of Earth</a>
  </li>
</ul>
</nav>

 <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        <div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    
    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/content/07/2/RNA_World.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
    
</div>
        <!-- Source interaction buttons -->


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/edur409/ASTRO200/master?urlpath=tree/content/07/2/RNA_World.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../../../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <button type="button" class="btn btn-secondary topbarbtn thebelab-launch-button"
            onclick="initThebelab()" title="Launch Thebelab" data-toggle="tooltip" data-placement="left"><i
                class="fas fa-rocket"></i><span style="margin-left: .4em;">ThebeLab</span></button>
        
    </div>
</div>

    </div>
    <div class="d-none d-md-block col-md-2 bd-toc show">
<div class="tocsection onthispage pt-5 pb-3">
    <i class="fas fa-list"></i> On this page
</div>

<nav id="bd-toc-nav">
    <ul class="nav section-nav flex-column">
    
        <li class="nav-item toc-entry toc-h2">
            <a href="#" class="nav-link">RNA World Hypothesis</a><ul class="nav section-nav flex-column">
                
        <li class="nav-item toc-entry toc-h3">
            <a href="#population-evolution-in-an-rna-world" class="nav-link">Population Evolution in an RNA World</a><ul class="nav section-nav flex-column">
                
        <li class="nav-item toc-entry toc-h4">
            <a href="#building-an-rna-population" class="nav-link">Building an RNA population</a><ul class="nav section-nav flex-column">
                
        <li class="nav-item toc-entry toc-h5">
            <a href="#rna-sequence-structure" class="nav-link">RNA sequence-structure</a>
        </li>
    
            </ul>
        </li>
    
        <li class="nav-item toc-entry toc-h4">
            <a href="#fitness-of-a-sequence-target-structure" class="nav-link">Fitness of a sequence: Target Structure</a>
        </li>
    
            </ul>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#defining-a-cell" class="nav-link">Defining a cell</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#populations-of-cells" class="nav-link">Populations of Cells</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#the-fitness-of-a-cell" class="nav-link">The Fitness of a Cell</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#introducing-diversity-mutations" class="nav-link">Introducing diversity: Mutations</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#selection" class="nav-link">Selection</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#gathering-information-on-our-populations" class="nav-link">Gathering information on our populations</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#and-finally-evolution" class="nav-link">And finally…. evolution</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#introducing-mating-to-the-model" class="nav-link">Introducing mating to the model</a>
        </li>
    
            </ul>
        </li>
    
        <li class="nav-item toc-entry toc-h2">
            <a href="#acknowledgements" class="nav-link">Acknowledgements</a>
        </li>
    
    </ul>
</nav>


    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="rna-world-hypothesis">
<h1>RNA World Hypothesis<a class="headerlink" href="#rna-world-hypothesis" title="Permalink to this headline">¶</a></h1>
<p>RNA is a simpler cousin of DNA.  As you may know, RNA is widely thought to be the first self replicating life-form to arise perhaps around 4 billion years ago. One of the strongest arguments for this theory is that RNA is able to carry information in its nucleotides like DNA, and like protein, it is able to adopt higher order structures to catalyze reactions, such as self replication. So it is likely, and there is growing evidence that this is the case, that the first form of replicating life was RNA. And because of this dual property of RNA as a hereditary information vessel as well as a structural/functional element we can use RNA molecules to build very nice population models.</p>
<p>So in this section, we’ll be walking you through building genetic populations, simulating their evolution, and using statistics and other mathematical tools for understanding key properties of populations.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">HTML</span>

<span class="c1"># Youtube</span>
<span class="n">HTML</span><span class="p">(</span><span class="s1">&#39;&lt;iframe width=&quot;560&quot; height=&quot;315&quot; src=&quot;https://www.youtube.com/embed/K1xnYFCZ9Yg&quot; frameborder=&quot;0&quot; allow=&quot;accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture&quot; allowfullscreen&gt;&lt;/iframe&gt;&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="stderr docutils container">
<pre class="stderr literal-block">/home/leo/anaconda3/lib/python3.6/site-packages/IPython/core/display.py:717: UserWarning: Consider using IPython.display.IFrame instead
  warnings.warn(&quot;Consider using IPython.display.IFrame instead&quot;)
</pre>
</div>
<div class="output text_html"><iframe width="560" height="315" src="https://www.youtube.com/embed/K1xnYFCZ9Yg" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#HTML(&#39;&lt;iframe width=&quot;784&quot; height=&quot;441&quot; src=&quot;https://scaleofuniverse.com/&quot; /iframe&gt;&#39;)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="population-evolution-in-an-rna-world">
<h2>Population Evolution in <em>an</em> RNA World<a class="headerlink" href="#population-evolution-in-an-rna-world" title="Permalink to this headline">¶</a></h2>
<p>In order to study the evolution of a population, we first need a model of a population. And even before that, we need to define what we mean by <em>population</em>. Populations can be defined on many levels and with many diffferent criteria. For our purposes, we will simply say that a population is a set of individuals sharing a common environment. And because this is population <em>genetics</em> we can think of individuals as entities comprising of specific genes or chromosomes.</p>
<p>So where do we get a population from? As you may have discussed in previous workshops, there are very large datasets containing sequencing information from different populations. So we could download one of these datasets and perform some analysis on it. But I find this can be dry and tedious. So why download data when we can simply create our own?</p>
<p>In this section we’re going to be creating and studying our own “artificial” populations to illustrate some important population genetics concepts and methodologies. Not only will this help you learn population genetics, but you will get a lot more programming practice than if we were to simply parse data files and go from there.</p>
<p>More specifically, we’re going to build our own RNA world.</p>
<div class="section" id="building-an-rna-population">
<h3>Building an RNA population<a class="headerlink" href="#building-an-rna-population" title="Permalink to this headline">¶</a></h3>
<p>As we saw earlier, RNA has the nice property of posessing a strong mapping between information carrying (sequence) and function (structure). This is analogous to what is known in evolutionary terms as a genotype and a phenotype. With these properties, we have everything we need to model a population, and simulate its evolution.</p>
<div class="section" id="rna-sequence-structure">
<h4>RNA sequence-structure<a class="headerlink" href="#rna-sequence-structure" title="Permalink to this headline">¶</a></h4>
<p>We can think of the genotype as a sequence <span class="math notranslate nohighlight">\(s\)</span> consisting of letters/nucleotides from the alphabet <span class="math notranslate nohighlight">\(\{U,A,C,G\}\)</span>. The corresponding phenotype <span class="math notranslate nohighlight">\(\omega\)</span> is the secondary structure of <span class="math notranslate nohighlight">\(s\)</span> which can be thought of as a pairing between nucleotides in the primary sequence that give rise to a 2D architecture. Because it has been shown that the function of many biomolecules, including RNA, is driven by structure this gives us a good proxy for phenotype.</p>
<p>Below is an example of what an RNA secondary structure, or pairing, looks like.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">### 1</span>

<span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">Image</span>
<span class="c1">#This will load an image of an RNA secondary structure</span>
<span class="n">Image</span><span class="p">(</span><span class="n">url</span><span class="o">=</span><span class="s1">&#39;https://viennarna.github.io/forgi/_images/1y26_ss.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><img src="https://viennarna.github.io/forgi/_images/1y26_ss.png"/></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">HTML</span>

<span class="c1"># Youtube</span>
<span class="n">HTML</span><span class="p">(</span><span class="s1">&#39;&lt;iframe width=&quot;560&quot; height=&quot;315&quot; src=&quot;https://www.youtube.com/embed/JQByjprj_mA&quot; frameborder=&quot;0&quot; allow=&quot;accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture&quot; allowfullscreen&gt;&lt;/iframe&gt;&#39;</span><span class="p">)</span>
<span class="c1">#import matplotlib.pyplot as plt</span>
<span class="c1">#import forgi.visual.mplotlib as fvm</span>
<span class="c1">#import forgi</span>
<span class="c1">#cg = forgi.load_rna(&quot;1y26.fx&quot;, allow_many=False)</span>
<span class="c1">#fvm.plot_rna(cg, text_kwargs={&quot;fontweight&quot;:&quot;black&quot;}, lighten=0.7,</span>
<span class="c1">#             backbone_kwargs={&quot;linewidth&quot;:3})</span>
<span class="c1">#plt.show()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><iframe width="560" height="315" src="https://www.youtube.com/embed/JQByjprj_mA" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe></div></div>
</div>
<p>As you can see, unpaired positions are forming loop-like structures, and paired positions are forming stem-like structures. It is this spatial arrangement of nucleotides that drives RNA’s function. Therefore, another sequence that adopts a similar shape, is likely to behave in a similar manner. Another thing to notice is that, although in reality this is often not the case, in general we only allow pairs between <span class="math notranslate nohighlight">\(\{C,G\}\)</span> and <span class="math notranslate nohighlight">\(\{A, U\}\)</span> nucleotides, most modern approaches allow for non-canonical pairings and you will find some examples of this in the above structure.</p>
<p><em>How do we go from a sequence to a structure?</em></p>
<p>So a secondary structure is just a list of pairings between positions. How do we get the optimal pairing?</p>
<p>The algorithm we’re going to be using in our simulations is known as the Nussinov Algorithm. The Nussinov algorithm is one of the first and simplest attempts at predicting RNA structure. Because bonds tend to stabilize RNA, the algorithm tries to maximize the number of pairs in the structure and return that as its solution. Current approaches achieve more accurate solutions by using energy models based one experimental values to then obtain a structure that minimizes free energy. But since we’re not really concerned with the accuracy of our predictions, Nussinov is a good entry point. Furthermore, the main algorithmic concepts are the same between Nussinov and state of the art RNA structure prediction algorithms. I implemented the algorithm in a separate file called <code class="docutils literal notranslate"><span class="pre">fold.py</span></code> that we can import and use its functions. I’m not going to go into detail here on how the algorithm works because it is beyond the scope of this workshop but there is a bonus exercise at the end if you’re curious.</p>
<p>You can predict a secondary structure by calling <code class="docutils literal notranslate"><span class="pre">nussinov()</span></code> with a sequence string and it will return a tuple in the form <code class="docutils literal notranslate"><span class="pre">(structure,</span> <span class="pre">pairs)</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">### 2</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">fold</span> <span class="kn">import</span> <span class="n">nussinov</span> <span class="c1">#Codes by Carlos G. Oliver (https://github.com/cgoliver)</span>

<span class="n">sequence_to_fold</span> <span class="o">=</span> <span class="s2">&quot;ACCCGAUGUUAUAUAUACCU&quot;</span>
<span class="n">struc</span> <span class="o">=</span> <span class="n">nussinov</span><span class="p">(</span><span class="n">sequence_to_fold</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&gt;test&quot;</span><span class="p">)</span> <span class="c1">#creates the structure of a .fx file for &quot;forna&quot;</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sequence_to_fold</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">struc</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="c1">#Check the molecule at: http://rna.tbi.univie.ac.at/forna/</span>
<span class="c1"># Paste the text below in the webpage and see the structure</span>
<span class="c1">#&gt;test</span>
<span class="c1">#ACCCGAUGUUAUAUAUACCU</span>
<span class="c1">#(...(..(((....).))))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;test
ACCCGAUGUUAUAUAUACCU
(...(..(((....).))))
</pre></div>
</div>
</div>
</div>
<p>You will see a funny dot-bracket string in the output. This is a representation of the structure of an RNA. Quite simply, a matching parir of parentheses (open and close) correspond to the nucleotides at those positions being paried. Whereas, a dot means that that position is unpaired in the structure. Feel free to play around with the input sequence to get a better understanding of the notation.</p>
<p>If you want to visually check the sequence, go to: <a class="reference external" href="http://rna.tbi.univie.ac.at/forna/forna.html">forna</a> and copy the text above with the sequence and its structure in the <strong>Add Molecule</strong> button.  The webpage is embedded below.</p>
<p>So that’s enough about RNA structure prediction. Let’s move on to building our populations.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">HTML</span><span class="p">(</span><span class="s1">&#39;&lt;iframe width=&quot;784&quot; height=&quot;441&quot; src=&quot;http://rna.tbi.univie.ac.at/forna/forna.html&quot; /iframe&gt;&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><iframe width="784" height="441" src="http://rna.tbi.univie.ac.at/forna/forna.html" /iframe></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span><span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>[2 2 2]
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="fitness-of-a-sequence-target-structure">
<h3>Fitness of a sequence: Target Structure<a class="headerlink" href="#fitness-of-a-sequence-target-structure" title="Permalink to this headline">¶</a></h3>
<p>Now that we have a good way of getting a phenotype (secondary structure), we need a way to evaluate the fitness of that phenotype. If we think in real life terms, fitness is the ability of a genotype to replicate into the next generation. If you have a gene carrying a mutation that causes some kind of disease, your fitness is decreased and you have a lower chance of contributing offspring to the next generation. On a molecular level the same concept applies. A molecule needs to accomplish a certain function, i.e. bind to some other molecule or send some kind of signal. And as we’ve seen before, the most important factor that determines how well it can carry out this function is its structure. So we can imagine that a certain structure, we can call this a ‘target’ structure, is required in order to accomplish a certain function. So a sequence that folds correctly to a target structure is seen as having a greater fitness than one that does not. Since we’ve encoded structures as simple dot-bracket strings, we can easily compare structures and thus evaluate the fitness between a given structure and the target, or ‘correct’ structure.</p>
<p>There are many ways to compare structures <span class="math notranslate nohighlight">\(w_{1}\)</span> and <span class="math notranslate nohighlight">\(w_{2}\)</span>, but we’re going to use one of the simplest ways, which is base-pair distance. This is just the number of pairs in <span class="math notranslate nohighlight">\(w_{1}\)</span> that are not in <span class="math notranslate nohighlight">\(w_{2}\)</span>. Again, this is beyond the scope of this workshop so I’ll just give you the code for it and if you would like to know more you can ask me.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">### 3</span>

<span class="c1">#ss_to_bp() and bp_distance() by Vladimir Reinharz.</span>
<span class="k">def</span> <span class="nf">ss_to_bp</span><span class="p">(</span><span class="n">ss</span><span class="p">):</span>
    <span class="n">bps</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">l</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ss</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="s1">&#39;(&#39;</span><span class="p">:</span>
                    <span class="n">l</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">x</span> <span class="o">==</span> <span class="s1">&#39;)&#39;</span><span class="p">:</span>
                    <span class="n">bps</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">l</span><span class="o">.</span><span class="n">pop</span><span class="p">(),</span> <span class="n">i</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">bps</span>

<span class="k">def</span> <span class="nf">bp_distance</span><span class="p">(</span><span class="n">w1</span><span class="p">,</span> <span class="n">w2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    return base pair distance between structures w1 and w1. </span>
<span class="sd">    w1 and w1 are lists of tuples representing pairing indices.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">w1</span><span class="p">)</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">w2</span><span class="p">)))</span>

<span class="c1">#let&#39;s fold two sequences</span>
<span class="n">w1</span> <span class="o">=</span> <span class="n">nussinov</span><span class="p">(</span><span class="s2">&quot;CCAAAAGG&quot;</span><span class="p">)</span>
<span class="n">w2</span> <span class="o">=</span> <span class="n">nussinov</span><span class="p">(</span><span class="s2">&quot;ACAAAAGA&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">w1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">w2</span><span class="p">)</span>

<span class="c1">#give the list of pairs to bp_distance and see what the distance is.</span>
<span class="nb">print</span><span class="p">(</span><span class="n">bp_distance</span><span class="p">(</span><span class="n">w1</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">w2</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>(&#39;((....))&#39;, [(0, 7), (1, 6)])
(&#39;.(....).&#39;, [(1, 6)])
1
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="defining-a-cell">
<h2>Defining a cell<a class="headerlink" href="#defining-a-cell" title="Permalink to this headline">¶</a></h2>
<p>The cell we will define here is a simple organism with two copies of an RNA gene, each with its own structure.  This simple definition of a cell will help us create populations to play around in our evolutionary reactor.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">### 4</span>
<span class="k">class</span> <span class="nc">Cell</span><span class="p">:</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_1</span><span class="p">,</span> <span class="n">struc_1</span><span class="p">,</span> <span class="n">seq_2</span><span class="p">,</span> <span class="n">struc_2</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence_1</span> <span class="o">=</span> <span class="n">seq_1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence_2</span> <span class="o">=</span> <span class="n">seq_2</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">structure_1</span> <span class="o">=</span> <span class="n">struc_1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">structure_2</span> <span class="o">=</span> <span class="n">struc_2</span>
        
<span class="c1">#for now just try initializing a Cell with made up sequences and structures</span>
<span class="n">cell</span> <span class="o">=</span> <span class="n">Cell</span><span class="p">(</span><span class="s2">&quot;AACCCCUU&quot;</span><span class="p">,</span> <span class="s2">&quot;((.....))&quot;</span><span class="p">,</span> <span class="s2">&quot;GGAAAACA&quot;</span><span class="p">,</span> <span class="s2">&quot;(....).&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cell</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">,</span> <span class="n">cell</span><span class="o">.</span><span class="n">structure_2</span><span class="p">,</span> <span class="n">cell</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">,</span> <span class="n">cell</span><span class="o">.</span><span class="n">structure_2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>AACCCCUU (....). AACCCCUU (....).
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="populations-of-cells">
<h2>Populations of Cells<a class="headerlink" href="#populations-of-cells" title="Permalink to this headline">¶</a></h2>
<p>Now we’ve defined a ‘Cell’. Since a population is a collection of individuals our populations will naturally consist of <strong>lists</strong> of ‘Cell’ objects, each with their own sequences. Here we initialize all the Cells with random sequences and add them to the ‘population’ list.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">### 5</span>
<span class="kn">import</span> <span class="nn">random</span>

<span class="k">def</span> <span class="nf">populate</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Creates a population of cells (pop_size) with a number of random RNA nucleotides (AUCG)</span>
<span class="sd">    matching the length of the target structure&#39;&#39;&#39;</span>
    <span class="n">population</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">pop_size</span><span class="p">):</span>
        <span class="c1">#get a random sequence to start with</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="s2">&quot;AUCG&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">target</span><span class="p">))])</span>
        <span class="c1">#use nussinov to get the secondary structure for the sequence</span>
        <span class="n">structure</span> <span class="o">=</span> <span class="n">nussinov</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="c1">#add a new Cell object to the population list</span>
        <span class="n">new_cell</span> <span class="o">=</span> <span class="n">Cell</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">structure</span><span class="p">,</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">structure</span><span class="p">)</span>
        <span class="n">new_cell</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">i</span>
        <span class="n">new_cell</span><span class="o">.</span><span class="n">parent</span> <span class="o">=</span> <span class="n">i</span>
        <span class="n">population</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_cell</span><span class="p">)</span>
            
    <span class="k">return</span> <span class="n">population</span>
</pre></div>
</div>
</div>
</div>
<p>Try creating a new population and printing the first 10 sequences and structures (in dot-bracket) on the first chromosome!</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">### 6</span>
<span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;(((....)))&quot;</span>
<span class="n">pop</span> <span class="o">=</span> <span class="n">populate</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">pop_size</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">[:</span><span class="mi">10</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">structure_1</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">p</span><span class="o">.</span><span class="n">sequence_2</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">structure_2</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    
<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">[</span><span class="o">-</span><span class="mi">10</span><span class="p">:]:</span><span class="c1">#for p in pop[:10]:#for p in pop[-10:]:#for p in pop[:10]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">structure_1</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">p</span><span class="o">.</span><span class="n">sequence_2</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">structure_2</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>0 GCAAUGUCUU ..((....)) GCAAUGUCUU ..((....))
1 AGGUGUUUAA (....).... AGGUGUUUAA (....)....
2 UGUCCCUGGG ...(....). UGUCCCUGGG ...(....).
3 GAGGUGAAGU .(.......) GAGGUGAAGU .(.......)
4 GGCGUUCCGA ((....)).. GGCGUUCCGA ((....))..
5 UGAGCACUGC .((....).) UGAGCACUGC .((....).)
6 GUCGCUAGAU .((....)). GUCGCUAGAU .((....)).
7 CAAUAUACCC .......... CAAUAUACCC ..........
8 GUGUUGUAAG .(.....).. GUGUUGUAAG .(.....)..
9 GCAUGUCCCA (.....)... GCAUGUCCCA (.....)...
290 GGAGCGGCAA (......).. GGAGCGGCAA (......)..
291 CUAUCUGAGA ((.....)). CUAUCUGAGA ((.....)).
292 CGACCCGUGG (.(....)). CGACCCGUGG (.(....)).
293 GAUAUUCGUA ..((....)) GAUAUUCGUA ..((....))
294 GUUGACUCAC ((......)) GUUGACUCAC ((......))
295 CGCUUAAAAU ...(....). CGCUUAAAAU ...(....).
296 AGCUGCGAGC .((.....)) AGCUGCGAGC .((.....))
297 CCCGUGCACU (....).... CCCGUGCACU (....)....
298 GUCACUCGCG (.(....)). GUCACUCGCG (.(....)).
299 UCGAUGGUUU .(....)... UCGAUGGUUU .(....)...
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="the-fitness-of-a-cell">
<h2>The Fitness of a Cell<a class="headerlink" href="#the-fitness-of-a-cell" title="Permalink to this headline">¶</a></h2>
<p>Once we are able to create populations of cells, we need a way of asssessing their individual fitness.  In our model, a <em>Cell</em> is an object that contains two sequences of RNA, something analogous to having two copies of a gene in each chromosome.</p>
<p>We could simply loop through each <em>Cell</em> in the population and check the base pair distance to the target structure we defined.  However, this approach of using base-pair distance is not the best for defining fitness. There are two reasons for this:</p>
<ol class="simple">
<li><p>We want fitness to represent a <em>probability</em> that a cell will reproduce (pass its genes to the next generation), and base pair distance is, in our case, an integer value.</p></li>
<li><p>We want this probability to be a <em>relative</em> measure of fitness. That is, we want the fitness to be proportional to how good a cell is with respect to all others in the population. This touches on an important principle in evolution where we only need to be ‘better’ than the rest of the population (the competition) and not good in some absolute measure. For instance, if you and I are being chased by a predator. In order to survive, I only need to be faster than you, and not necessarily some absolute level of fitness.</p></li>
</ol>
<p>In order to get a probability (number between 0 and 1) we use the following equation to define the fitness of a structure <span class="math notranslate nohighlight">\(\omega\)</span> on a target structure <span class="math notranslate nohighlight">\(T\)</span>:</p>
<div class="math notranslate nohighlight">
\[P(\omega, T) = N^{-1} exp(\frac{-\beta \texttt{dist}(\omega, T)}{\texttt{len}(\omega)})\]</div>
<div class="math notranslate nohighlight">
\[N = \sum_{i \in Pop}{P(\omega_i, T})\]</div>
<p>Here, the <span class="math notranslate nohighlight">\(N\)</span> is what gives us the ‘relative’ measure because we divide the fitness of the Cell by the sum of the fitness of every other Cell.</p>
<p>Let’s take a quick look at how this function behaves if we plot different base pair distance values.</p>
<p>What is the effect of the parameter <span class="math notranslate nohighlight">\(\beta\)</span>? Try plotting the same function but with different values of <span class="math notranslate nohighlight">\(\beta\)</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">matplotlib</span> <span class="n">inline</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">math</span>

<span class="n">target_length</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">beta</span> <span class="o">=</span> <span class="o">-</span><span class="mi">3</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">beta</span> <span class="o">*</span> <span class="p">(</span><span class="n">bp_dist</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">target_length</span><span class="p">)))</span> <span class="k">for</span> <span class="n">bp_dist</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">target_length</span><span class="p">)])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Base pair distance to target structure&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;P(w, T)&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>Text(0, 0.5, &#39;P(w, T)&#39;)
</pre></div>
</div>
<img alt="../../../_images/RNA_World_20_1.png" src="../../../_images/RNA_World_20_1.png" />
</div>
</div>
<p>As you can see, it’s a very simple function that evaluates to 1 (highest fitness) if the base pair distance is 0, and decreases as the structures get further and further away from the target. I didn’t include the <span class="math notranslate nohighlight">\(N\)</span> in the plotting as it will be a bit more annoying to compute, but it is simply a scaling factor so the shape and main idea won’t be different.</p>
<p>Now we can use this function to get a fitness value for each Cell in our population.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">### 7</span>

<span class="k">def</span> <span class="nf">compute_fitness</span><span class="p">(</span><span class="n">population</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">beta</span><span class="o">=-</span><span class="mi">3</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Assigns a fitness and bp_distance value to each cell in the population.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#store the fitness values of each cell</span>
    <span class="n">tot</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1">#iterate through each cell</span>
    <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">population</span><span class="p">:</span>
        
        <span class="c1">#calculate the bp_distance of each chromosome using the cell&#39;s structure</span>
        <span class="n">bp_distance_1</span> <span class="o">=</span> <span class="n">bp_distance</span><span class="p">(</span><span class="n">cell</span><span class="o">.</span><span class="n">structure_1</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">ss_to_bp</span><span class="p">(</span><span class="n">target</span><span class="p">))</span>
        <span class="n">bp_distance_2</span> <span class="o">=</span> <span class="n">bp_distance</span><span class="p">(</span><span class="n">cell</span><span class="o">.</span><span class="n">structure_2</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">ss_to_bp</span><span class="p">(</span><span class="n">target</span><span class="p">))</span>
        
        <span class="c1">#use the bp_distances and the above fitness equation to calculate the fitness of each chromosome</span>
        <span class="n">fitness_1</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">((</span><span class="n">beta</span> <span class="o">*</span> <span class="n">bp_distance_1</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cell</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">))))</span>
        <span class="n">fitness_2</span> <span class="o">=</span>  <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">((</span><span class="n">beta</span> <span class="o">*</span> <span class="n">bp_distance_2</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cell</span><span class="o">.</span><span class="n">sequence_2</span><span class="p">))))</span>

        <span class="c1">#get the fitness of the whole cell by multiplying the fitnesses of each chromosome</span>
        <span class="n">cell</span><span class="o">.</span><span class="n">fitness</span> <span class="o">=</span> <span class="n">fitness_1</span> <span class="o">*</span> <span class="n">fitness_2</span>
               
        <span class="c1">#store the bp_distance of each chromosome.</span>
        <span class="n">cell</span><span class="o">.</span><span class="n">bp_distance_1</span> <span class="o">=</span> <span class="n">bp_distance_1</span>
        <span class="n">cell</span><span class="o">.</span><span class="n">bp_distance_2</span> <span class="o">=</span> <span class="n">bp_distance_2</span>
    
        
        <span class="c1">#add the cell&#39;s fitness value to the list of all fitness values (used for normalization later)</span>
        <span class="n">tot</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cell</span><span class="o">.</span><span class="n">fitness</span><span class="p">)</span>

    <span class="c1">#normalization factor is sum of all fitness values in population</span>
    <span class="n">norm</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">tot</span><span class="p">)</span>
    <span class="c1">#divide all fitness values by the normalization factor.</span>
    <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">population</span><span class="p">:</span>
        <span class="n">cell</span><span class="o">.</span><span class="n">fitness</span> <span class="o">=</span> <span class="n">cell</span><span class="o">.</span><span class="n">fitness</span> <span class="o">/</span> <span class="n">norm</span>

    <span class="k">return</span> <span class="kc">None</span>

<span class="n">compute_fitness</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span>
<span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">[:</span><span class="mi">10</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">cell</span><span class="o">.</span><span class="n">fitness</span><span class="p">,</span> <span class="n">cell</span><span class="o">.</span><span class="n">bp_distance_1</span><span class="p">,</span> <span class="n">cell</span><span class="o">.</span><span class="n">bp_distance_2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>0.001080618002141078 5 5
0.0019690143777416906 4 4
0.0019690143777416906 4 4
0.0019690143777416906 4 4
0.001080618002141078 5 5
0.0035877781159223522 3 3
0.011911842837697651 1 1
0.0035877781159223522 3 3
0.0019690143777416906 4 4
0.0019690143777416906 4 4
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="introducing-diversity-mutations">
<h2>Introducing diversity: Mutations<a class="headerlink" href="#introducing-diversity-mutations" title="Permalink to this headline">¶</a></h2>
<p>Evolution would go nowhere without random mutations. While mutations are technically just random errors in the copying of genetic material, they are essential in the process of evolution. This is because they introduce novel diversity to populatons, which with a low frequency can be beneficial. And when a beneficial mutation arises (i.e. a mutation that increases fitness, or replication probability) it quickly takes over the population and the populatioin as a whole has a higher fitness.</p>
<p>Implementing mutations in our model will be quite straightforward. Since mutations happen at the genotype/sequence level, we simply have to iterate through our strings of nucleotides (sequences) and randomly introduce changes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">mutate</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">0.001</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Takes a sequence and mutates bases with probability mutation_rate&quot;&quot;&quot;</span>
    
    <span class="c1">#start an empty string to store the mutated sequence</span>
    <span class="n">new_sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="c1">#boolean storing whether or not the sequence got mutated</span>
    <span class="n">mutated</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="c1">#go through every bp in the sequence</span>
    <span class="k">for</span> <span class="n">bp</span> <span class="ow">in</span> <span class="n">sequence</span><span class="p">:</span>
        <span class="c1">#generate a random number between 0 and 1</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span>
        <span class="c1">#if r is below mutation rate, introduce a mutation</span>
        <span class="k">if</span> <span class="n">r</span> <span class="o">&lt;</span> <span class="n">mutation_rate</span><span class="p">:</span>
            <span class="c1">#add a randomly sampled nucleotide to the new sequence</span>
            <span class="n">new_sequence</span> <span class="o">=</span> <span class="n">new_sequence</span> <span class="o">+</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="s2">&quot;aucg&quot;</span><span class="p">)</span>
            <span class="n">mutated</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">#if the mutation condition did not get met, copy the current bp to the new sequence</span>
            <span class="n">new_sequence</span> <span class="o">=</span> <span class="n">new_sequence</span> <span class="o">+</span> <span class="n">bp</span>
            
    <span class="k">return</span> <span class="p">(</span><span class="n">new_sequence</span><span class="p">,</span> <span class="n">mutated</span><span class="p">)</span>

<span class="n">sequence_to_mutate</span> <span class="o">=</span> <span class="s1">&#39;AAAAGGAGUGUGUAUGU&#39;</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sequence_to_mutate</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mutate</span><span class="p">(</span><span class="n">sequence_to_mutate</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">0.5</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>AAAAGGAGUGUGUAUGU
(&#39;gAAAGgAGUGUGUAUGU&#39;, True)
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="selection">
<h2>Selection<a class="headerlink" href="#selection" title="Permalink to this headline">¶</a></h2>
<p>The final process in this evolution model is selection. Once you have populations with a diverse range of fitnesses, we need to select the fittest individuals and let them replicate and contribute offspring to the next generation. In real populations this is just the process of reproduction. If you’re fit enough you will be likely to reproduce more than another individual who is not as well suited to the environment.</p>
<p>In order to represent this process in our model, we will use the fitness values that we assigned to each Cell earlier and use that to select replicating Cells. This is equivalent to sampling from a population with the sampling being weighted by the fitness of each Cell. Thankfully, <code class="docutils literal notranslate"><span class="pre">numpy.random.choice</span></code> comes to the rescue here. Once we have sampled enough Cells to build our next generation, we introduce mutations and compute the fitness values of the new generation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">selection</span><span class="p">(</span><span class="n">population</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">beta</span><span class="o">=-</span><span class="mi">3</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a new population with offspring of the input population</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">#select the sequences that will be &#39;parents&#39; and contribute to the next generation</span>
    <span class="n">parents</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">population</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">population</span><span class="p">),</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="n">rna</span><span class="o">.</span><span class="n">fitness</span> <span class="k">for</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">population</span><span class="p">],</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1">#build the next generation using the parents list</span>
    <span class="n">next_generation</span> <span class="o">=</span> <span class="p">[]</span>    
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">p</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">parents</span><span class="p">):</span>
        <span class="n">new_cell</span> <span class="o">=</span> <span class="n">Cell</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">structure_1</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">sequence_2</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">structure_2</span><span class="p">)</span>
        <span class="n">new_cell</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">i</span>
        <span class="n">new_cell</span><span class="o">.</span><span class="n">parent</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">id</span>
        
        <span class="n">next_generation</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_cell</span><span class="p">)</span>

    <span class="c1">#introduce mutations in next_generation sequeneces and re-fold when a mutation occurs</span>
    <span class="k">for</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">next_generation</span><span class="p">:</span>      
        <span class="n">mutated_sequence_1</span><span class="p">,</span> <span class="n">mutated_1</span> <span class="o">=</span> <span class="n">mutate</span><span class="p">(</span><span class="n">rna</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">)</span>
        <span class="n">mutated_sequence_2</span><span class="p">,</span> <span class="n">mutated_2</span> <span class="o">=</span> <span class="n">mutate</span><span class="p">(</span><span class="n">rna</span><span class="o">.</span><span class="n">sequence_2</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">mutated_1</span><span class="p">:</span>
            <span class="n">rna</span><span class="o">.</span><span class="n">sequence_1</span> <span class="o">=</span> <span class="n">mutated_sequence_1</span>
            <span class="n">rna</span><span class="o">.</span><span class="n">structure_1</span> <span class="o">=</span> <span class="n">nussinov</span><span class="p">(</span><span class="n">mutated_sequence_1</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mutated_2</span><span class="p">:</span>
            <span class="n">rna</span><span class="o">.</span><span class="n">sequence_2</span> <span class="o">=</span> <span class="n">mutated_sequence_2</span>
            <span class="n">rna</span><span class="o">.</span><span class="n">structure_2</span> <span class="o">=</span> <span class="n">nussinov</span><span class="p">(</span><span class="n">mutated_sequence_2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">continue</span>

    <span class="c1">#update fitness values for the new generation</span>
    <span class="n">compute_fitness</span><span class="p">(</span><span class="n">next_generation</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="n">beta</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">next_generation</span>

<span class="n">next_gen</span> <span class="o">=</span> <span class="n">selection</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span>
<span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">next_gen</span><span class="p">[:</span><span class="mi">10</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">cell</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>UGAGCACUGC
UGAGCACUGC
CGGAACUUAG
GUCUAUCGGC
AUAACGGUAU
CCAGCCUUUG
GGCGCUCGAC
UUGGUAGACA
AUGGGAUGAU
CGUAGGGGCG
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="gathering-information-on-our-populations">
<h2>Gathering information on our populations<a class="headerlink" href="#gathering-information-on-our-populations" title="Permalink to this headline">¶</a></h2>
<p>Here we simply store some statistics (in a dictionary) on the population at each generation such as the average base pair distance and the average fitness of the populations. No coding to do here, it’s not a very interesting function but feel free to give it a look.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">record_stats</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">population_stats</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Takes a population list and a dictionary and updates it with stats on the population.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">generation_bp_distance_1</span> <span class="o">=</span> <span class="p">[</span><span class="n">rna</span><span class="o">.</span><span class="n">bp_distance_1</span> <span class="k">for</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">]</span>
    <span class="n">generation_bp_distance_2</span> <span class="o">=</span> <span class="p">[</span><span class="n">rna</span><span class="o">.</span><span class="n">bp_distance_2</span> <span class="k">for</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">]</span>

    <span class="n">mean_bp_distance_1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">generation_bp_distance_1</span><span class="p">)</span>
    <span class="n">mean_bp_distance_2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">generation_bp_distance_2</span><span class="p">)</span>
    
    <span class="n">mean_fitness</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">rna</span><span class="o">.</span><span class="n">fitness</span> <span class="k">for</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">])</span>


    <span class="n">population_stats</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s1">&#39;mean_bp_distance_1&#39;</span><span class="p">,</span> <span class="p">[])</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mean_bp_distance_1</span><span class="p">)</span>
    <span class="n">population_stats</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s1">&#39;mean_bp_distance_2&#39;</span><span class="p">,</span> <span class="p">[])</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mean_bp_distance_2</span><span class="p">)</span>
    
    <span class="n">population_stats</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s1">&#39;mean_fitness&#39;</span><span class="p">,</span> <span class="p">[])</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mean_fitness</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="kc">None</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="and-finally-evolution">
<h2>And finally…. evolution<a class="headerlink" href="#and-finally-evolution" title="Permalink to this headline">¶</a></h2>
<p>We can put all the above parts together in a simple function that does the following:</p>
<ol class="simple">
<li><p>start a new population and compute its fitness</p></li>
<li><p>repeat the following for the desired number of generations:</p>
<ol class="simple">
<li><p>record statistics on population</p></li>
<li><p>perform selection+mutation</p></li>
<li><p>store new population</p></li>
</ol>
</li>
</ol>
<p>And that’s it! We have an evolutionary reactor!</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">evolve</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">generations</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">pop_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">beta</span><span class="o">=-</span><span class="mi">2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Takes target structure and sets up initial population, performs selection and iterates for desired generations.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#store list of all populations throughotu generations [[cells from generation 1], [cells from gen. 2]...]</span>
    <span class="n">populations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1">#start a dictionary that will hold some stats on the populations.</span>
    <span class="n">population_stats</span> <span class="o">=</span> <span class="p">{}</span>
    
    <span class="c1">#get a starting population</span>
    <span class="n">initial_population</span> <span class="o">=</span> <span class="n">populate</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">pop_size</span><span class="o">=</span><span class="n">pop_size</span><span class="p">)</span>
    <span class="c1">#compute fitness of initial population</span>
    <span class="n">compute_fitness</span><span class="p">(</span><span class="n">initial_population</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span>

    <span class="c1">#set current_generation to initial population.</span>
    <span class="n">current_generation</span> <span class="o">=</span> <span class="n">initial_population</span>

    <span class="c1">#iterate the selection process over the desired number of generations</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">generations</span><span class="p">):</span>

        <span class="c1">#let&#39;s get some stats on the structures in the populations   </span>
        <span class="n">record_stats</span><span class="p">(</span><span class="n">current_generation</span><span class="p">,</span> <span class="n">population_stats</span><span class="p">)</span>
        
        <span class="c1">#add the current generation to our list of populations.</span>
        <span class="n">populations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_generation</span><span class="p">)</span>

        <span class="c1">#select the next generation</span>
        <span class="n">new_gen</span> <span class="o">=</span> <span class="n">selection</span><span class="p">(</span><span class="n">current_generation</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="n">beta</span><span class="p">)</span>
        <span class="c1">#set current generation to be the generation we just obtained.</span>
        <span class="n">current_generation</span> <span class="o">=</span> <span class="n">new_gen</span> 
    
    <span class="k">return</span> <span class="p">(</span><span class="n">populations</span><span class="p">,</span> <span class="n">population_stats</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Try a run of the <code class="docutils literal notranslate"><span class="pre">evolve()</span></code> function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;(((....)))&quot;</span>
<span class="n">pops</span><span class="p">,</span> <span class="n">pops_stats</span> <span class="o">=</span> <span class="n">evolve</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">generations</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">pop_size</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">beta</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>

<span class="c1">#Print the first 10 sequences of the population</span>
<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">[:</span><span class="mi">10</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">structure_1</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">p</span><span class="o">.</span><span class="n">sequence_2</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">structure_2</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

<span class="c1">#Print the last 10 sequences of the population    </span>
<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">[</span><span class="o">-</span><span class="mi">10</span><span class="p">:]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">structure_1</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">p</span><span class="o">.</span><span class="n">sequence_2</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">structure_2</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>0 GCAAUGUCUU ..((....)) GCAAUGUCUU ..((....))
1 AGGUGUUUAA (....).... AGGUGUUUAA (....)....
2 UGUCCCUGGG ...(....). UGUCCCUGGG ...(....).
3 GAGGUGAAGU .(.......) GAGGUGAAGU .(.......)
4 GGCGUUCCGA ((....)).. GGCGUUCCGA ((....))..
5 UGAGCACUGC .((....).) UGAGCACUGC .((....).)
6 GUCGCUAGAU .((....)). GUCGCUAGAU .((....)).
7 CAAUAUACCC .......... CAAUAUACCC ..........
8 GUGUUGUAAG .(.....).. GUGUUGUAAG .(.....)..
9 GCAUGUCCCA (.....)... GCAUGUCCCA (.....)...
290 GGAGCGGCAA (......).. GGAGCGGCAA (......)..
291 CUAUCUGAGA ((.....)). CUAUCUGAGA ((.....)).
292 CGACCCGUGG (.(....)). CGACCCGUGG (.(....)).
293 GAUAUUCGUA ..((....)) GAUAUUCGUA ..((....))
294 GUUGACUCAC ((......)) GUUGACUCAC ((......))
295 CGCUUAAAAU ...(....). CGCUUAAAAU ...(....).
296 AGCUGCGAGC .((.....)) AGCUGCGAGC .((.....))
297 CCCGUGCACU (....).... CCCGUGCACU (....)....
298 GUCACUCGCG (.(....)). GUCACUCGCG (.(....)).
299 UCGAUGGUUU .(....)... UCGAUGGUUU .(....)...
</pre></div>
</div>
</div>
</div>
<p>Let’s see if it actually worked by plotting the average base pair distance as a function of generations for both genes in each cell. We should expect a gradual decrease as the populations get closer to the target structure.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">evo_plot</span><span class="p">(</span><span class="n">pops_stats</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot base pair distance for each chromosome over generations.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="s1">&#39;Mean base pair Distance&#39;</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;mean_bp_distance_1&#39;</span><span class="p">,</span> <span class="s1">&#39;mean_bp_distance_2&#39;</span><span class="p">]:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">pops_stats</span><span class="p">[</span><span class="n">m</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">m</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Generations&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Mean Base Pair Distance&quot;</span><span class="p">)</span>
    
<span class="n">evo_plot</span><span class="p">(</span><span class="n">pops_stats</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/RNA_World_34_0.png" src="../../../_images/RNA_World_34_0.png" />
</div>
</div>
<p>Let’s see the structure of random cells from each generation.  Run the code below and copy the output in the RNA folding webpage. Compare the base-pair distance plot with the structures.</p>
<p>Questions:</p>
<ul class="simple">
<li><p>Do you notice some simmilarities from a particular generation onwards? Compare your observations to the plot with the Mean Base Pair Distance.</p></li>
<li><p>What could trigger another evolutionary jump?</p></li>
</ul>
<p><img alt="Structure of RNA" src="content/07/2/RNA_evolution.png" /></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">HTML</span>
<span class="n">HTML</span><span class="p">(</span><span class="s1">&#39;&lt;iframe width=&quot;784&quot; height=&quot;441&quot; src=&quot;http://rna.tbi.univie.ac.at/forna/forna.html&quot; /iframe&gt;&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><iframe width="784" height="441" src="http://rna.tbi.univie.ac.at/forna/forna.html" /iframe></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#Select a random RNA sequence from each generation to check its folding structure</span>
<span class="kn">from</span> <span class="nn">random</span> <span class="kn">import</span> <span class="n">randrange</span>
<span class="c1">#print(randrange(999))</span>
<span class="n">generations</span><span class="o">=</span><span class="mi">20</span>
<span class="n">pop_size</span><span class="o">=</span><span class="mi">300</span>
<span class="c1">#Print some random cells from each generation </span>
<span class="c1">#pops[generation][cell in that generation].{quality to retrieve}</span>
<span class="k">for</span> <span class="n">gen</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">generations</span><span class="p">):</span>
    <span class="n">cid</span><span class="o">=</span><span class="n">randrange</span><span class="p">(</span><span class="n">pop_size</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&gt;Gen&#39;</span><span class="o">+</span><span class="n">np</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">gen</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_Cell_&#39;</span><span class="o">+</span><span class="n">np</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">pops</span><span class="p">[</span><span class="n">gen</span><span class="p">][</span><span class="n">cid</span><span class="p">]</span><span class="o">.</span><span class="n">id</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">pops</span><span class="p">[</span><span class="n">gen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">+</span><span class="n">np</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">pops</span><span class="p">[</span><span class="n">gen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">structure_1</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Gen1_Cell_176
UUCAGGGUGU
..(.....).

&gt;Gen2_Cell_279
UGGCGCUGUA
(........)

&gt;Gen3_Cell_240
GCAUCACUGC
(((....)))

&gt;Gen4_Cell_2
GCAUCACUGC
(((....)))

&gt;Gen5_Cell_277
UGCUCCAACA
((......))

&gt;Gen6_Cell_164
GUAAUAGUAC
(((....)))

&gt;Gen7_Cell_238
GCAUCACUGC
(((....)))

&gt;Gen8_Cell_79
CACGACUGUG
(((....)))

&gt;Gen9_Cell_56
GGCAUUAGCC
(((....)))

&gt;Gen10_Cell_41
GCAgCACUGC
(((....)))

&gt;Gen11_Cell_299
GCAgCACUGC
(((....)))

&gt;Gen12_Cell_227
GCAUCACUGC
(((....)))

&gt;Gen13_Cell_190
CACGACUGUG
(((....)))

&gt;Gen14_Cell_240
CCUAAUUAGG
(((....)))

&gt;Gen15_Cell_15
GUAAUAGUAC
(((....)))

&gt;Gen16_Cell_219
CCUAAUUAGG
(((....)))

&gt;Gen17_Cell_202
CACGACUGUG
(((....)))

&gt;Gen18_Cell_193
GUAAUAGUAC
(((....)))

&gt;Gen19_Cell_157
GUAAUAGUAC
(((....)))

&gt;Gen20_Cell_172
CACGACUGUG
(((....)))

</pre></div>
</div>
</div>
</div>
<p>You should see a nice drop in base pair distance! Another way of visualizing this is by plotting a histogram of the base pair distance of all Cells in the initial population versus the final population.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#import seaborn as sns</span>
<span class="c1">#def bp_distance_distributions(pops):</span>
<span class="c1">#    &quot;&quot;&quot;</span>
<span class="c1">#    Plots histograms of base pair distance in initial and final populations.</span>
<span class="c1">#    &quot;&quot;&quot;</span>
<span class="c1">#    #plot bp_distance_1 for rnas in first population</span>
<span class="c1">#   g = sns.distplot([rna.bp_distance_1 for rna in pops[0]], label=&#39;initial population&#39;)</span>
<span class="c1">#    #plot bp_distance_1 for rnas in first population</span>
<span class="c1">#    g = sns.distplot([rna.bp_distance_1 for rna in pops[-1]], label=&#39;final population&#39;)</span>
<span class="c1">#    g.set(xlabel=&#39;Mean Base Pair Distance&#39;)</span>
<span class="c1">#    g.legend()</span>
<span class="c1">#</span>
<span class="c1">#bp_distance_distributions(pops)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="introducing-mating-to-the-model">
<h2>Introducing mating to the model<a class="headerlink" href="#introducing-mating-to-the-model" title="Permalink to this headline">¶</a></h2>
<p>The populations we generated evolved asexually. This means that individuals do not mate or exchange genetic information. So to make our simulation a bit more interesting let’s let the Cells mate. This is going to require a few small changes in the <code class="docutils literal notranslate"><span class="pre">selection()</span></code> function. Previously, when we selected sequences to go into the next generation we just let them provide one offspring which was a copy of itself and introduced mutations. Now instead of choosing one Cell at a time, we will randomly choose two ‘parents’ that will mate. When they mate, each parent will contribute one of its chromosomes to the child. We’ll repeat this process until we have filled the next generation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">selection_with_mating</span><span class="p">(</span><span class="n">population</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">beta</span><span class="o">=-</span><span class="mi">2</span><span class="p">):</span>
    <span class="n">next_generation</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">next_generation</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">population</span><span class="p">):</span>
        <span class="c1">#select two parents based on their fitness</span>
        <span class="n">parents_pair</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">population</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="n">rna</span><span class="o">.</span><span class="n">fitness</span> <span class="k">for</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">population</span><span class="p">],</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        
        <span class="c1">#take the sequence and structure from the first parent&#39;s first chromosome and give it to the child</span>
        <span class="n">child_chrom_1</span> <span class="o">=</span> <span class="p">(</span><span class="n">parents_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">,</span> <span class="n">parents_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">structure_1</span><span class="p">)</span>

        <span class="c1">#do the same for the child&#39;s second chromosome and the second parent.</span>
        <span class="n">child_chrom_2</span> <span class="o">=</span> <span class="p">(</span><span class="n">parents_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sequence_2</span><span class="p">,</span> <span class="n">parents_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">structure_2</span><span class="p">)</span>


        <span class="c1">#initialize the new child Cell with the new chromosomes.</span>
        <span class="n">child_cell</span> <span class="o">=</span> <span class="n">Cell</span><span class="p">(</span><span class="n">child_chrom_1</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">child_chrom_1</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">child_chrom_2</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">child_chrom_2</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

        <span class="c1">#give the child and id and store who its parents are</span>
        <span class="n">child_cell</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">counter</span>
        <span class="n">child_cell</span><span class="o">.</span><span class="n">parent_1</span> <span class="o">=</span> <span class="n">parents_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
        <span class="n">child_cell</span><span class="o">.</span><span class="n">parent_2</span> <span class="o">=</span> <span class="n">parents_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>

        <span class="c1">#add the child to the new generation</span>
        <span class="n">next_generation</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">child_cell</span><span class="p">)</span>
        
        <span class="n">counter</span> <span class="o">=</span> <span class="n">counter</span> <span class="o">+</span> <span class="mi">1</span>
            
        
    <span class="c1">#introduce mutations in next_generation sequeneces and re-fold when a mutation occurs (same as before)</span>
    <span class="k">for</span> <span class="n">rna</span> <span class="ow">in</span> <span class="n">next_generation</span><span class="p">:</span>      
        <span class="n">mutated_sequence_1</span><span class="p">,</span> <span class="n">mutated_1</span> <span class="o">=</span> <span class="n">mutate</span><span class="p">(</span><span class="n">rna</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">)</span>
        <span class="n">mutated_sequence_2</span><span class="p">,</span> <span class="n">mutated_2</span> <span class="o">=</span> <span class="n">mutate</span><span class="p">(</span><span class="n">rna</span><span class="o">.</span><span class="n">sequence_2</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">mutated_1</span><span class="p">:</span>
            <span class="n">rna</span><span class="o">.</span><span class="n">sequence_1</span> <span class="o">=</span> <span class="n">mutated_sequence_1</span>
            <span class="n">rna</span><span class="o">.</span><span class="n">structure_1</span> <span class="o">=</span> <span class="n">nussinov</span><span class="p">(</span><span class="n">mutated_sequence_1</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mutated_2</span><span class="p">:</span>
            <span class="n">rna</span><span class="o">.</span><span class="n">sequence_2</span> <span class="o">=</span> <span class="n">mutated_sequence_2</span>
            <span class="n">rna</span><span class="o">.</span><span class="n">structure_2</span> <span class="o">=</span> <span class="n">nussinov</span><span class="p">(</span><span class="n">mutated_sequence_2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">continue</span>

    <span class="c1">#update fitness values for the new generation</span>
    <span class="n">compute_fitness</span><span class="p">(</span><span class="n">next_generation</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="n">beta</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">next_generation</span>    

<span class="c1">#run a small test to make sure it works</span>
<span class="n">next_gen</span> <span class="o">=</span> <span class="n">selection_with_mating</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span>
<span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">next_gen</span><span class="p">[:</span><span class="mi">10</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">cell</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>GUGGGUUCUU
GGGUAAAACC
AAGAACGUAG
CUGGAUCCCG
GGCCAUAGAA
ACACACGAGU
GACGCCGGUG
AUCGUCUGCU
AUUGUUGGUU
UAGGGAACUU
</pre></div>
</div>
</div>
</div>
<p>Now we just have to update our <code class="docutils literal notranslate"><span class="pre">evolution()</span></code> function to call the new <code class="docutils literal notranslate"><span class="pre">selection_with_mating()</span></code> function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">evolve_with_mating</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">generations</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">pop_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">beta</span><span class="o">=-</span><span class="mi">2</span><span class="p">):</span>
    <span class="n">populations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">population_stats</span> <span class="o">=</span> <span class="p">{}</span>
    
    <span class="n">initial_population</span> <span class="o">=</span> <span class="n">populate</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">pop_size</span><span class="o">=</span><span class="n">pop_size</span><span class="p">)</span>
    <span class="n">compute_fitness</span><span class="p">(</span><span class="n">initial_population</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span>
        
    <span class="n">current_generation</span> <span class="o">=</span> <span class="n">initial_population</span>

    <span class="c1">#iterate the selection process over the desired number of generations</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">generations</span><span class="p">):</span>
        <span class="c1">#let&#39;s get some stats on the structures in the populations   </span>
        <span class="n">record_stats</span><span class="p">(</span><span class="n">current_generation</span><span class="p">,</span> <span class="n">population_stats</span><span class="p">)</span>
        
        <span class="c1">#add the current generation to our list of populations.</span>
        <span class="n">populations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_generation</span><span class="p">)</span>

        <span class="c1">#select the next generation, but this time with mutations</span>
        <span class="n">new_gen</span> <span class="o">=</span> <span class="n">selection_with_mating</span><span class="p">(</span><span class="n">current_generation</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="n">beta</span><span class="p">)</span>
        <span class="n">current_generation</span> <span class="o">=</span> <span class="n">new_gen</span> 
    
    <span class="k">return</span> <span class="p">(</span><span class="n">populations</span><span class="p">,</span> <span class="n">population_stats</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Try out the new evolution model!</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pops_mating</span><span class="p">,</span> <span class="n">pops_stats_mating</span> <span class="o">=</span> <span class="n">evolve_with_mating</span><span class="p">(</span><span class="s2">&quot;(((....)))&quot;</span><span class="p">,</span> <span class="n">generations</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">pop_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">beta</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>

<span class="n">evo_plot</span><span class="p">(</span><span class="n">pops_stats_mating</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/RNA_World_45_0.png" src="../../../_images/RNA_World_45_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#Select a random RNA sequence from each generation to check its folding structure</span>
<span class="kn">from</span> <span class="nn">random</span> <span class="kn">import</span> <span class="n">randrange</span>
<span class="c1">#print(randrange(999))</span>
<span class="n">generations</span><span class="o">=</span><span class="mi">20</span>
<span class="n">pop_size</span><span class="o">=</span><span class="mi">1000</span>
<span class="c1">#Print some random cells from each generation </span>
<span class="c1">#pops[generation][cell in that generation].{quality to retrieve}</span>
<span class="k">for</span> <span class="n">gen</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">generations</span><span class="p">):</span>
    <span class="n">cid</span><span class="o">=</span><span class="n">randrange</span><span class="p">(</span><span class="n">pop_size</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&gt;Gen&#39;</span><span class="o">+</span><span class="n">np</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">gen</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_Cell_&#39;</span><span class="o">+</span><span class="n">np</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">pops_mating</span><span class="p">[</span><span class="n">gen</span><span class="p">][</span><span class="n">cid</span><span class="p">]</span><span class="o">.</span><span class="n">id</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">pops_mating</span><span class="p">[</span><span class="n">gen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">sequence_1</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">+</span><span class="n">np</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">pops_mating</span><span class="p">[</span><span class="n">gen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">structure_1</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Gen1_Cell_315
GUCAGAGCUG
..((....))

&gt;Gen2_Cell_44
GGUCGGUGCC
((......))

&gt;Gen3_Cell_839
AGGACCAACU
((......))

&gt;Gen4_Cell_742
UCUACCAUGG
(.....)...

&gt;Gen5_Cell_368
GGUUCUGACC
(((....)))

&gt;Gen6_Cell_906
GGUGCAGACC
(((....)))

&gt;Gen7_Cell_603
UGUAGCGACA
(((....)))

&gt;Gen8_Cell_656
UGUAGCGACA
(((....)))

&gt;Gen9_Cell_36
AAUAACUAUU
(((....)))

&gt;Gen10_Cell_970
GUUCACAAAC
(((....)))

&gt;Gen11_Cell_105
AGGUAUCCCU
(((....)))

&gt;Gen12_Cell_238
AUUUACAAAU
(((....)))

&gt;Gen13_Cell_362
UGUAGCGACA
(((....)))

&gt;Gen14_Cell_823
AAGUGGCCAU
(.(....).)

&gt;Gen15_Cell_475
AAUAACUAUU
(((....)))

&gt;Gen16_Cell_624
UGUUGGAACA
(((....)))

&gt;Gen17_Cell_186
AAUAACUAUU
(((....)))

&gt;Gen18_Cell_181
GGCUUUGGCC
(((....)))

&gt;Gen19_Cell_0
UGUAGCGACA
(((....)))

&gt;Gen20_Cell_378
AAAGUACUUU
(((....)))

</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="acknowledgements">
<h1>Acknowledgements<a class="headerlink" href="#acknowledgements" title="Permalink to this headline">¶</a></h1>
<p>The computational codes of this notebook were originally created by <a class="reference external" href="https://github.com/cgoliver">Carlos G. Oliver</a>, and adapted by Evert Durán for ASTRO200.</p>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "edur409/ASTRO200",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "content/07/2"
        }
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="../1/complex-life.html" title="previous page">Complex Life</a>
    <a class='right-next' id="next-link" href="../../08-habitability-our-system.html" title="next page">Habitability of Inner vs. Outer Solar System</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Evert Duran<br/>
        
            &copy; Copyright 2020.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    <script src="../../../_static/js/index.js"></script>
    
  </body>
</html>